Introduce CoreWCF project templates

g7ed6e opened the pull request:
Resolves #541

g7ed6e commented on src/CoreWCF.Templates/src/CoreWCF.Templates.csproj:
```diff
@@ -0,0 +1,26 @@
+  <PropertyGroup>
+    <PackageType>Template</PackageType>
+    <PackageVersion>1.0</PackageVersion>
+    <!--<PackageId>AdatumCorporation.Utility.Templates</PackageId>
```
clean

g7ed6e commented on azure-pipelines-ci.yml:
```diff
@@ -14,7 +14,7 @@ pr: none
   _solution: 'src/CoreWCF.sln'
   _libraryProjects: 'src/CoreWCF.*/src/CoreWCF.*.csproj'
-  _testProjects: '**/*.Tests.csproj'
+  _libraryTestProjects: '**/*.Tests.csproj'
```
'**/*.Tests.csproj;!**/*.Templates.Tests.csproj'

samsp-msft commented on src/CoreWCF.Templates/src/templates/.editorconfig:
```diff
@@ -0,0 +1,133 @@
+# Pour en savoir plus sur .editorconfig, consultez https://aka.ms/editorconfigdocs
```
Is there one of these for the project as a whole - if so why is this different, and if not, should this be moved up to apply at the root of the repo?

g7ed6e replied:
This file overrides is there to override the one from CoreWCF repo as i guess template users does not want the MIT license header.

samsp-msft commented on src/CoreWCF.Templates/src/templates/CoreWCFService/IService.cs:
```diff
@@ -0,0 +1,61 @@
+        }
+    }
+
+    // Utilisez un contrat de données comme indiqué dans l'exemple ci-après pour ajouter les types composites aux opérations de service.
```
Can we have the comment in English please - I don't think we have localization support for templates. I'm assuming this is the french version of "// Use a data contract as illustrated in the sample below to add composite types to service operations."

g7ed6e replied:
Sure

samsp-msft replied:
I took it from the .NET Framework project, so I'm assuming its the same.

samsp-msft commented on src/CoreWCF.Templates/src/templates/CoreWCFService/Program.cs:
```diff
@@ -0,0 +1,39 @@
+﻿var builder = WebApplication.CreateBuilder();
```
Are the \#if's resolved at template creation time or is this the code expected to be seen by the user?

If the goal is to enable all the different scenarios, then I think we should reconsider. There is a balance needed in templates between:
* The least number of lines
* The simplest form of the code
* Handling every possibility

Generally we favor the simplest form of the code - so as to not to scare users off, and to provide a good starting point.

I think we should pick a default for the template, and have that for both the modern form and classic program.cs formats. I would suggest:

| Feature | Included in template |
| --- | --- |
| WSDL | Yes |
| NetTCP | No |
| BasicHttpBinding | Yes |
| Https BasicHttpBinding | Yes |

g7ed6e replied:
> Are the #if's resolved at template creation time or is this the code expected to be seen by the user?

Yes please see https://github.com/dotnet/templating/wiki/Conditional-processing-and-comment-syntax
> 
> If the goal is to enable all the different scenarios, then I think we should reconsider. There is a balance needed in templates between:
> 
> * The least number of lines
> * The simplest form of the code
> * Handling every possibility
> 
> Generally we favor the simplest form of the code - so as to not to scare users off, and to provide a good starting point.
> 
> I think we should pick a default for the template, and have that for both the modern form and classic program.cs formats. I would suggest:
> 
> Feature	Included in template
> WSDL	Yes
> NetTCP	No
> BasicHttpBinding	Yes
> Https BasicHttpBinding	Yes

Then i will drop netTcp binding support

samsp-msft commented on src/CoreWCFTemplates.sln:
```diff
@@ -0,0 +1,31 @@
+﻿
```
I understand the need for this while you are developing the templates, but should these be incorporated into the main sln so that as people add support, the templates are also under their noses, so they know to update them too?

g7ed6e replied:
Running unit test requires the current code to be built and packaged thus i moved them in a separate solution to avoid embarassing users who want to work on CoreWCF runtime code. Please see https://github.com/dotnet/aspnetcore/blob/main/src/ProjectTemplates/README.md#running-projecttemplate-tests , i'm trying to do something similar but without the hive dependency.

mconnew commented on src/CoreWCF.Templates/src/templates/CoreWCFService/IService.cs:
```diff
@@ -0,0 +1,61 @@
+﻿#if(!impliciteUsings)
+using System.Runtime.Serialization;
+using CoreWCF;
+using System;
```
Order of usings should be alphabetical. So:
```c#
using CoreWCF;
using System;
using System.Runtime.Serialization;
```

g7ed6e commented on README.md:
````diff
@@ -33,6 +35,17 @@ To keep up to date on what's going on with CoreWCF, you can subscribe to the [an
+dotnet new corewcf
+```
+* `CoreWCFService` project template supports the following arguments
+  * `--binding`: `nettcp` or `basic` (default)
````
this is no longer true.
add supported options and default values

g7ed6e replied:
done

g7ed6e commented on templates/TestTemplatesStage.yml:
```diff
@@ -0,0 +1,56 @@
+    - task: PowerShell@2
+      inputs:
+        targetType: 'filePath'
+        filePath: src/CoreWCF.Templates/Run-Templates-UnitTests.ps1
```
@mconnew  I ensured the script return a non zero exit code to fail the build unit tests are failing

g7ed6e commented on src/CoreWCF.Templates/src/templates/CoreWCFService/.template.config/dotnetcli.host.json:
```diff
@@ -0,0 +1,33 @@
+{
```
add the vs support file like here https://github.com/dotnet/aspnetcore/blob/main/src/ProjectTemplates/Web.ProjectTemplates/content/EmptyWeb-CSharp/.template.config/ide.host.json

g7ed6e replied:
![image](https://user-images.githubusercontent.com/681739/165898921-f0795a80-d6de-4bf5-af1f-8d14c2180ba3.png)

g7ed6e replied:
![image](https://user-images.githubusercontent.com/681739/165899131-dac9a0fd-3d8b-4989-9ba9-67f3fc948c22.png)

g7ed6e commented on src/CoreWCF.Templates/src/templates/CoreWCFService/Program.cs:
```diff
@@ -0,0 +1,34 @@
+
+#if (!NoWsdl)
+    var serviceMetadataBehavior = app.Services.GetRequiredService<ServiceMetadataBehavior>();
+
```
drop this line

g7ed6e commented on src/CoreWCF.Templates/tests/BasicTests.cs:
```diff
@@ -0,0 +1,212 @@
+            return this;
+        }
+
+        public TestVariation Net6()
```
drop

g7ed6e commented on README.md:
```diff
@@ -29,6 +29,19 @@ To keep up to date on what's going on with CoreWCF, you can subscribe to the [an
+It supports the following arguments:
+  * `--framework`: `net6.0` (default), `net5.0`, `netcoreapp3.1`, `net48`, `net472` and `net462` are valid values.
+  * `--use-program-main`: whether to turn off ASP.NET Core minimal API hosting. This option only affects `net6.0` projects as other target require a `Startup` and a `Program` class. 
+  * `--no-https`: whether to turn off HTTPS and use BasicHttpSecurityMode.None. Default is HTTPS enabled a `BasicHttpBinding` using `BasicHttpSecurityMode.Transport`.
```
Missing ˋ around BasicHttpSecurityMode.None

g7ed6e replied:
Missing `using` a ..

g7ed6e commented on src/CoreWCF.Templates/src/templates/CoreWCFService/.template.config/template.json:
```diff
@@ -0,0 +1,189 @@
+{
+  "$schema": "http://json.schemastore.org/template",
+  "author": "CoreWCF",
```
@mconnew author is displayed in VS new project wizard.

samsp-msft replied:
I would suggest either:
* CoreWCF Project 
* or if it doesn't balk at having a URL as the author, then https://github.com/CoreWCF/CoreWCF

g7ed6e replied:
I used https://github.com/CoreWCF as author.

g7ed6e commented on src/CoreWCF.Templates/src/templates/CoreWCFService/.template.config/ide.host.json:
```diff
@@ -0,0 +1,27 @@
+{
+    "$schema": "http://json.schemastore.org/vs-2017.3.host",
+    "icon": "ide/Empty.png",
```
This is a default empty C# project logo

g7ed6e commented on src/CoreWCF.Templates/src/templates/CoreWCFService/.template.config/template.json:
```diff
@@ -0,0 +1,189 @@
+  },
+  "name": "Core WCF Service",
+  "defaultName": "CoreWCFService",
+  "description": "A project template to host a Core WCF service.",
```
Is this wording correct?

samsp-msft replied:
```suggestion
  "description": "A project template for creating a server application which hosts a Core WCF service.",
```

Kralizek commented:
As someone who is extremely n00b on GitHub and is just lurking the repo. Is it normal that whenever I land on this PR from the notification summary, I see nothing in the "files changed" panel?

g7ed6e commented:
> As someone who is extremely n00b on GitHub and is just lurking the repo. Is it normal that whenever I land on this PR from the notification summary, I see nothing in the "files changed" panel?

i already encountered the issue you are describing with other PR. But with this one it works for me. I see all modifications there https://github.com/CoreWCF/CoreWCF/pull/590/files

Kralizek commented:
@g7ed6e 
If I click the link you posted, it works. Same is if i click again on the "modified files".

Could it be that the notification points to commits that have been removed from the PR path via forced pushes?

g7ed6e commented:
> @g7ed6e If I click the link you posted, it works. Same is if i click again on the "modified files".
> 
> Could it be that the notification points to commits that have been removed from the PR path via forced pushes?

yes i force pushed several times

Kralizek commented:
Oh okay. In my noobness i thought it was very bad to force push. Sorry for the annoyance, I'll get back to lurk.

mconnew commented:
@Kralizek, it's bad to force push to a branch other people will be working from. The model we use here is make your own fork of the repo, make your branch there, then make a pull request from that branch in your fork to the CoreWCF repo. When you force push to your own branch, GitHub updates the PR and reruns the tests. The problem with force pushing comes if I were to force push to main. That changes the history of the branch and when other people are making changes based off of main, it makes it very hard to reconcile their changes when the base branch has had its history changed. But it's perfectly normal and common to force push to a temporary branch used for feature development.

mconnew commented:
@g7ed6e, just an FYI, I'm going to take my time reviewing this and will have multiple people try it out and give their thoughts on it. I want to make sure the initial release has an intuitive user experience. It might take a few weeks to fully review this.

Kralizek commented:
@mconnew thanks for the explanation

samsp-msft commented on src/CoreWCF.Templates/src/templates/CoreWCFService/IService.cs:
```diff
@@ -0,0 +1,61 @@
+using System.Runtime.Serialization;
+#endif
+
+namespace CoreWCFService
```
Its not usual for templates to have multiple classes etc in one file. But in this case, I think its appropriate. It makes it very clear that this is all the contract stuff, and so you don't have to go fishing to find it.

samsp-msft commented on src/CoreWCF.Templates/src/templates/CoreWCFService/Properties/launchSettings.json:
```diff
@@ -0,0 +1,16 @@
+  "profiles": {
+    "CoreWCFService": {
+      "commandName": "Project",
+      "launchBrowser": true,
```
IMHO this line should be removed. We don't have the fancy test page that used to exist for soap services. The browser comes up saying it can't find a url, and it controls the lifetime of the process - so closing the browser stops debugging. Without it the process stays running and you can then test with a client, query for wsdl etc.

g7ed6e replied:
i dropped the launchBrowser

samsp-msft commented:
This is freaking cool, I love it. I created a couple of variations to see how they look and they all look good. The main change I'd ask for is to remove the launch browser from launch.settings, but otherwise is good.

@mconnew & @g7ed6e- Is there a way that we could publish a preview of the template package (that refs 1.0 of corewcf) and ask for user feedback. I think this is one place where people really need to try it to find issues. That and a quick post in the announcement repo asking people to try it?

samsp-msft commented:
@g7ed6e - For the <PackageReference ... > in the resulting csproj it seems to be picking up a version from the current build. Should this actually be a wildcard so to start with its would be 
```xml
<PackageReference Include="CoreWCF.Http" Version="1.*" />
```
That means it would pick up the latest version of 1.x but wouldn't pickup a 2.0. The goals for API compat that we have stated for incremental releases is that they won't break existing APIs.

g7ed6e commented:
1. Add a separate pipeline to release the templates indenpendently
2. Exclude modifications done to files matching the templates folder from triggers

g7ed6e commented:
> @g7ed6e - For the <PackageReference ... > in the resulting csproj it seems to be picking up a version from the current build. Should this actually be a wildcard so to start with its would be
> 
> ```
> <PackageReference Include="CoreWCF.Http" Version="1.*" />
> ```
> 
> That means it would pick up the latest version of 1.x but wouldn't pickup a 2.0. The goals for API compat that we have stated for incremental releases is that they won't break existing APIs.

done

g7ed6e commented:
@mconnew i had a look to the path exclusion you requested to be added in the TestTemplates stage I introduced in the PR pipeline and I think this is not natively supported by azure devops. 
https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml
https://stackoverflow.com/questions/61213779/azure-devops-condition-on-sources-paths
Thus i added a new pipeline. azure-pipelines-templates.yml. However the new pipeline is not ran in my PR, I think a new build definition should be created in azure devops.

mconnew commented on src/CoreWCF.Templates/src/templates/CoreWCFService/CoreWCFService.csproj:
```diff
@@ -0,0 +1,28 @@
+<Project Sdk="Microsoft.NET.Sdk.Web">
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <!--#if(nullableEnabled)-->
```
We have support for targeting .NET Framework, but this looks hard coded to .net6.0. Should there be an if block around this?

g7ed6e replied:
I double checked this is correctly valued dynamically using the optional `--framework` arg when using dotnet CLI. Using the VS GUI you may also use the drop down menu allowing to choose the target framework. https://github.com/CoreWCF/CoreWCF/pull/590#discussion_r861507711

mconnew replied:
Sounds like it's just a placeholder then. That's really non-intuitive. Thanks for checking.

g7ed6e replied:
indeed. You can have a look to the integration tests here:
https://github.com/CoreWCF/CoreWCF/blob/3f3e41801daeb787f1cfc86bfc202fcce12fd6b4/src/CoreWCF.Templates/tests/BasicTests.cs
Basically CoreWCF template is installed and then the unit test code is issuing all variations of `dotnet new corewcf`. Then the project is built and ran. Finally an HttpClient ensure it is working appropriately.

mconnew commented on src/CoreWCF.Templates/src/CoreWCF.Templates.csproj:
```diff
@@ -0,0 +1,22 @@
+    <Content Remove="templates\Directory.Build.props"/>
+    <Content Remove="templates\Directory.Build.targets"/>
+    <Content Remove="templates\.editorconfig"/>
+    <Compile Remove="templates\**\*"/>
```
I'm having difficulty understanding some of these. You are removing the .editorconfig, but I thought we had a separate one so that the generated project included an editorconfig?

mconnew replied:
I see my confusion now. I didn't notice there was separate Content and Compile items. The general convention is to not mix different item types in a single ItermGroup. That always confused me why, but I think I just learnt why. Can you move the Compile into it's own ItemGroup to make things clearer.

g7ed6e replied:
done. i also added comments

mconnew commented on README.md:
```diff
@@ -29,6 +29,20 @@ To keep up to date on what's going on with CoreWCF, you can subscribe to the [an
 * The [Blog](https://corewcf.github.io/) has details on the design philosophy and a deep dive into the features are included in each new release.
 
+### Use Core WCF project templates (dotnet CLI or VisualStudio)
+
```
There was some ambiguity with the name where it was sometimes referred to as "Core WCF" and other times "CoreWCF". We settled on "CoreWCF" and the rest of the documentation has been updated to reflect that. Can you fix the few references here?

mconnew commented on src/CoreWCF.Templates/src/templates/CoreWCFService/Program.Main.cs:
```diff
@@ -0,0 +1,32 @@
+                    webBuilder.UseKestrel(options =>
+                    {
+                        options.AllowSynchronousIO = true;
+                    });
```
This isn't needed any more unless you are using streamed transfer mode.

g7ed6e replied:
it dropped it in [32d1a91](https://github.com/CoreWCF/CoreWCF/pull/590/commits/32d1a917a5c8ef67a8b585e09aa7a6b0febd0058)

mconnew replied:
@FreeAndNil, it depends on what your contract looks like. If you are using data types which will use a serializer, then you will need to turn on `AllowSynchronousIO`. This is because we create an XmlWriter which writes directly to the network and pass that to the serializer and tell it to serialize. The serializers only have a synchronous implementation so write to the XmlWriter using synchronous api's only, which means the output stream only gets written to synchronously. If you are using `Stream` in your contract, I believe this doesn't need this option, but I haven't tested it. If it does still need it, I might be able to improve that.

FreeAndNil replied:
We use a stream as response.
We get no error, but after some time  (12 to 30 hours in our development environment) the server does no longer respond to new requests. We see many, many threads when looking at the dump.

I would propose opening a new issue.

mconnew replied:
Please do open a new issue. Are you able to provide a process dump? You can contact me privately on gitter to arrange making a dump available.

mconnew commented:
/azp run
